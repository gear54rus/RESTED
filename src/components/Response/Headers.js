import React, { PropTypes } from 'react';
import Highlight from 'react-highlight';

import Collapse from 'components/Collapsible';
import { responseShape } from 'propTypes/response';

function Headers({ headers, expanded }) {
  if (expanded) {
    return (
      <div>
        <h4>Headers</h4>
        <Highlight className="http">
          {headers.reduce((prev, header) => (
            `${prev ? `${prev}\n` : ''}${header.name}: ${header.value}`
          ), '')}
        </Highlight>
      </div>
    );
  }

  return (
    <Collapse
      title="Headers"
      id="Headers"
    >
      <Highlight className="http">
        {headers.reduce((prev, header) => (
          `${prev ? `${prev}\n` : ''}${header.name}: ${header.value}`
        ), '')}
      </Highlight>
    </Collapse>
  );
}

Headers.defaultProps = {
  expanded: false,
};

Headers.propTypes = {
  headers: responseShape.headers,
  expanded: PropTypes.bool,
};

export default Headers;

