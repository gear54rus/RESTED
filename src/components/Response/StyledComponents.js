import React from 'react';
import styled from 'styled-components';
import { Panel } from 'react-bootstrap';
import Fonticon from 'components/Fonticon';

// Fix "Unknown prop" warnings
// eslint-disable-next-line
const PrunedPanel = ({ wrapResponse, ...rest }) => <Panel {...rest} />;

export const StyledResponse = styled(PrunedPanel)`
  pre {
    padding: 0;
  }

  pre, code {
    ${props => props.wrapResponse && 'white-space: pre-wrap;'};
  }
`;

export const LoadingSpinner = styled(Fonticon)`
  padding: 40px 0 40px;
  font-size: 5em;
  opacity: 0.9;
`;

export const StyledHeader = styled.div`
  ${props => props.expandable && 'cursor: pointer;'}

  h3 {
    font-size: 15px;
    margin: 4px 0;
  }
`;

export const Status = styled.span`
  font-size: 32px;
  color: orange;
  ${props => props.green && 'color: green;'}
  ${props => props.red && 'color: red;'}
`;

export const PreviewContainer = styled(Panel)`
  .panel-body {
    padding: 0;
  }

  iframe {
    border: 0;
    width: 100%;
    height: 600px;
  }
`;
