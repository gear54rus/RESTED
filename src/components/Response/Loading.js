import React from 'react';
import { Col } from 'react-bootstrap';

import { LoadingSpinner } from './StyledComponents';

export default function Loading() {
  return (
    <Col xs={12} className="text-center">
      <LoadingSpinner
        icon="circle-notch"
        className="fa-spin fa-9x"
      />
    </Col>
  );
}
