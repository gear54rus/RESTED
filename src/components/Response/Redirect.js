import React from 'react';
import PropTypes from 'prop-types';
import { Panel } from 'react-bootstrap';
import responsePropTypes, { redirectShape } from 'propTypes/redirect';

import { StyledResponse, StyledHeader, Status } from './StyledComponents';
import Headers from './Headers';

function Titlebar({ url, time, onClick }) {
  return ( // if we don't pass onClick, collapse won't work
    <StyledHeader expandable onClick={onClick}>
      <h3>
        Redirect ({(time / 1000).toFixed(3)}s) -{' '}
        <a
          className="text-muted"
          href={url}
          target="_blank"
          rel="noopener noreferrer"
        >
          {url}
        </a>
      </h3>
    </StyledHeader>
  );
}

Titlebar.propTypes = {
  url: redirectShape.url,
  time: redirectShape.time,
  onClick: PropTypes.func.isRequired,
};

function Redirect(props) {
  const {
    response,
    headers,
    eventKey,
  } = props;

  if (!response || !headers) return null;

  const { method, url, time } = response;

  return (
    <StyledResponse
      eventKey={eventKey}
    >
      <Panel.Heading>
        <Panel.Toggle // can't use title, collapse won't work
          componentClass={Titlebar}
          method={method}
          url={url}
          time={time}
        />
      </Panel.Heading>
      <Panel.Body collapsible>
        <h3>
          <Status
            green={response.statusCode >= 200 && response.statusCode < 300}
            red={response.statusCode >= 400 && response.statusCode < 600}
          >
            {response.statusCode}
          </Status>
          <small> {response.statusLine && response.statusLine.replace(/.*\d{3} /, '')}</small>
        </h3>

        <Headers expanded headers={headers} />
      </Panel.Body>
    </StyledResponse>
  );
}

Redirect.propTypes = {
  response: responsePropTypes,
  headers: redirectShape.responseHeaders,
  eventKey: PropTypes.number.isRequired,
};

export default Redirect;

