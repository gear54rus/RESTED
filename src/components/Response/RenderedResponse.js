import React from 'react';
import PropTypes from 'prop-types';

import Collapse from 'components/Collapsible';
import base64 from 'utils/base64';
import { PreviewContainer } from './StyledComponents';

/**
 * Preview the content of the request by rendering it
 * in an iframe.
 *
 * Uses the sandbox attribute to place restrictions on the
 * markup in the frame among other thins preventing
 * scripts from running. This should be suficcient as
 * we have control of the content outside of the iframe.
 */
function RenderedResponse({ html }) {
  return (
    <Collapse
      title="Preview"
      id="Preview"
      mountOnEnter
      unmountOnExit
    >
      <PreviewContainer>
        <iframe
          title="Response preview"
          src={`data:text/html;base64,${base64(html)}`}
          sandbox=""
          referrerPolicy="no-referrer"
        />
      </PreviewContainer>
    </Collapse>
  );
}

RenderedResponse.propTypes = {
  html: PropTypes.string.isRequired,
};

export default RenderedResponse;

