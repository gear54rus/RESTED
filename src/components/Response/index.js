import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { PanelGroup, Panel, Alert } from 'react-bootstrap';

import { getResponse, getInterceptedResponse, getRedirectChain, getLoading } from 'store/request/selectors';
import responsePropTypes from 'propTypes/response';

import Loading from './Loading';
import Redirect from './Redirect';
import Response from './Response';

function ResponseAccordion({ response, error, loading, redirectChain, interceptedResponse }) {
  if (error) {
    return (
      <Alert bsStyle="danger">
        {`An error occurred while fetching the resource: ${error}`}
      </Alert>
    );
  }

  if (loading) {
    return (
      <Panel>
        <Panel.Body>
          <Loading />
        </Panel.Body>
      </Panel>
    );
  }

  if (!response) return null;

  return (
    <PanelGroup accordion id="response-accordion">
      {redirectChain.map((redirectResponse, i) => (
        <Redirect
          response={redirectResponse}
          headers={redirectResponse.responseHeaders}
          key={i}
          eventKey={i}
        />
      ))}
      <Response
        response={response}
        redirectChain={redirectChain}
        interceptedResponse={interceptedResponse}
      />
    </PanelGroup>
  );
}

ResponseAccordion.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Error),
  ]),
  response: responsePropTypes,
  redirectChain: PropTypes.arrayOf(PropTypes.shape({})),
  interceptedResponse: PropTypes.shape({}),
};

const mapStateToProps = state => ({
  response: getResponse(state),
  interceptedResponse: getInterceptedResponse(state),
  redirectChain: getRedirectChain(state),
  error: state.request.error,
  loading: getLoading(state),
});

export default connect(mapStateToProps)(ResponseAccordion);
