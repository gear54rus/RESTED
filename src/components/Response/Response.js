import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Alert, Panel } from 'react-bootstrap';
import Highlight from 'react-highlight';
import formatXml from 'xml-formatter';

import * as Actions from 'store/request/actions';
import { isDisabledHighlighting, isWrapResponse } from 'store/options/selectors';
import responsePropTypes, { responseShape } from 'propTypes/response';
import getContentType from 'utils/contentType';
import approximateSizeFromLength from 'utils/approximateSizeFromLength';

import { StyledResponse, StyledHeader, Status } from './StyledComponents';
import Headers from './Headers';
import RenderedResponse from './RenderedResponse';

function Titlebar({ url, time }) {
  return (
    <StyledHeader>
      <h3>
        Response ({time}) -{' '}
        <a
          className="text-muted"
          href={url}
          target="_blank"
          rel="noopener noreferrer"
        >
          {url}
        </a>
      </h3>
    </StyledHeader>
  );
}

Titlebar.propTypes = {
  url: responseShape.url, // eslint-disable-line react/no-typos
  time: PropTypes.node.isRequired,
};

export function Response(props) {
  const {
    response,
    highlightingDisabled,
    wrapResponse,
    redirectChain,
    interceptedResponse,
  } = props;

  if (!response || !interceptedResponse) return null;

  const { method, url, totalTime } = response;
  let { body } = response;

  let time;
  if (redirectChain.length > 0) {
    time = `${(interceptedResponse.time / 1000).toFixed(3)}s, total time ${totalTime / 1000}s`;
  } else {
    time = `${totalTime / 1000}s`;
  }

  const contentLength = interceptedResponse.responseHeaders.find(header => (
    header.name.toLowerCase() === 'content-length'
  ));
  const contentType = interceptedResponse.responseHeaders.find(header => (
    header.name.toLowerCase() === 'content-type'
  ));

  const contentSize = contentLength
    ? Number(contentLength.value)
    : approximateSizeFromLength(body);
  const type = getContentType(contentType && contentType.value);

  try {
    if (type.json) {
      body = JSON.stringify(JSON.parse(body), null, 2);
    } else if (type.xml) {
      body = formatXml(body);
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('Encountered an error while formatting response as ' +
      `${contentType && contentType.value}. Falling back to plain text`, e);
  }

  return (
    <StyledResponse
      wrapResponse={wrapResponse}
    >
      <Panel.Heading>
        <Panel.Title
          componentClass={Titlebar}
          method={method}
          url={url}
          time={time}
        />
      </Panel.Heading>
      <Panel.Body>
        <h3>
          <Status
            green={response.status >= 200 && response.status < 300}
            red={response.status >= 400 && response.status < 600}
          >
            {response.status}
          </Status>
          <small> {response.statusText}</small>
        </h3>

        <Headers headers={interceptedResponse.responseHeaders} />
        {type.html && <RenderedResponse html={body} />}

        {!highlightingDisabled && contentSize < 20000
          ? (
            <Highlight>
              {body}
            </Highlight>
          ) : (
            <span>
              {contentSize >= 20000 && (
                <Alert bsStyle="warning">
                  The size of the response is greater than 20KB, syntax
                  highlighting has been disabled for performance reasons.
                </Alert>
              )}

              <code>
                <pre>
                  {body}
                </pre>
              </code>
            </span>
          )
        }
      </Panel.Body>
    </StyledResponse>
  );
}

Response.propTypes = {
  response: responsePropTypes,
  highlightingDisabled: PropTypes.bool.isRequired,
  wrapResponse: PropTypes.bool.isRequired,
  redirectChain: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  interceptedResponse: PropTypes.shape({}),
};

const mapStateToProps = state => ({
  highlightingDisabled: isDisabledHighlighting(state),
  wrapResponse: isWrapResponse(state),
});

export default connect(mapStateToProps, Actions)(Response);

