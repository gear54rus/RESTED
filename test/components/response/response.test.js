import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

/* eslint-disable import/no-unresolved */
import { Response } from 'components/Response/Response';
import makeStore from '../../makeStore';

jest.mock('react-highlight');

describe('response component', () => {
  const response = {
    url: 'http://example.com',
    method: 'GET',
    status: 200,
    statusText: 'OK',
    body: 'Some long body',
    totalTime: 123,
    headers: [{
      name: 'Content-Type',
      value: 'application/awesome',
    }],
  };

  const interceptedResponse = {
    ...response,
    responseHeaders: response.headers,
  };

  const store = makeStore({
    request: { response },
  });

  it('renders nothing given no props', () => {
    const tree = renderer.create(
      <Response loading={false} />,
    );

    expect(tree).toMatchSnapshot();
  });

  const props = {
    response,
    interceptedResponse,
    redirectChain: [],
  };

  // TODO move to test super component
  xit('renders a loading gif when requets is in flight', () => {
    const tree = mount(
      <Provider store={store}>
        <Response {...props} loading />
      </Provider>,
    );

    expect(!tree.find('Loading').isEmpty()).toBe(true);
  });

  it('renders a result when given one', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Response {...props} loading={false} />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('displays the status and statusText', () => {
    const tree = mount(
      <Provider store={store}>
        <Response {...props} loading={false} />
      </Provider>,
    );

    const h3 = tree.find('.panel-body h3');
    expect(h3.text()).toEqual('200 OK');

    const small = h3.find('small');
    expect(small.prop('children')).toContain('OK');
  });

  it('displays the URL and method in the titlebar', () => {
    const tree = mount(
      <Provider store={store}>
        <Response {...props} loading={false} />
      </Provider>,
    );

    const expectedLink = (
      <a
        href="http://example.com"
        className="text-muted"
        target="_blank"
        rel="noopener noreferrer"
      >
        http://example.com
      </a>
    );

    const heading = tree.find('.panel-heading');
    expect(heading.contains(expectedLink)).toEqual(true);
  });
});

