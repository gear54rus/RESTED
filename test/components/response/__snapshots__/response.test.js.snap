// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`response component renders a result when given one 1`] = `
<div
  className="sc-bdVaJa dwrHNK panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="sc-htpNat gBkAzt"
    >
      <h3>
        Response (
        0.123s
        ) -
         
        <a
          className="text-muted"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          http://example.com
        </a>
      </h3>
    </div>
  </div>
  <div
    className="panel-body"
  >
    <h3>
      <span
        className="sc-bxivhb hPXgMN"
      >
        200
      </span>
      <small>
         
        OK
      </small>
    </h3>
    <div>
      <button
        className="sc-gzVnrw JHKFV btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <h4
          className="sc-EHOje LDHyT"
        >
          Headers
          <i
            className="fas fa-angle-right sc-bZQynM jDdTPr"
            role="presentation"
          />
        </h4>
      </button>
      <div
        aria-expanded={null}
        className="collapse row"
      >
        <div
          className="col-xs-12"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`response component renders nothing given no props 1`] = `null`;
